src/io/video.js

```diff
@@ -0,0 +1,332 @@
+            } else if (format === Video.FORMAT_CANVAS) {
+                // this will never change
+                formatCache.lastUpdate = Infinity;
+                formatCache.lastData = canvas;
+            } else {
+                log.error(`video io error - unimplemented format ${format}`);
+                // cache the null result forever, don't log about it again..
+                formatCache.lastUpdate = Infinity;
+                formatCache.lastData = null;
+            }
+
+            // rather than set to now, this data is as stale as it's canvas is
+            formatCache.lastUpdate = Math.max(workspace.lastUpdate, formatCache.lastUpdate);
+        }
+
+        return formatCache.lastData;
+    }
+
+    /**
+     * Set the preview ghost effect
+     * @param {number} ghost from 0 (visible) to 100 (invisible) - ghost effect
+     */
+    setPreviewGhost (ghost) {
+        if (this._drawable) {
+            this.runtime.renderer.updateDrawableProperties(this._drawable, {ghost});
+        }
+    }
+
+
+    /**
+     * Create a video stream.
+     * Should probably be moved to -render or somewhere similar later
+     * @private
+     * @return {Promise} When video has been received, rejected if video is not received
+     */
+    _setupVideo () {
+        if (this._singleSetup) {
+            return this._singleSetup;
+        }
+
+        this._video = document.createElement('video');
+        const video = new Promise((resolve, reject) => {
+            navigator.getUserMedia({
+                audio: false,
+                video: {
+                    width: {min: 480, ideal: 640},
+                    height: {min: 360, ideal: 480}
+                }
+            }, stream => {
+                this._video.src = window.URL.createObjectURL(stream);
+                // Hint to the stream that it should load. A standard way to do this
+                // is add the video tag to the DOM. Since this extension wants to
+                // hide the video tag and instead render a sample of the stream into
+                // the webgl rendered Scratch canvas, another hint like this one is
+                // needed.
+                this._track = stream.getTracks()[0];
+                resolve(this._video);
+            }, err => {
+                // There are probably some error types we could handle gracefully here.
+                this._singleSetup = null;
+                reject(err);
+            });
+        });
+
+        this._singleSetup = video.then(() => this._setupPreview());
+        return this._singleSetup;
+    }
+
+    _disablePreview () {
+        if (this._skin && this._drawable) {
+            this._skin.clear();
+            this.runtime.renderer.updateDrawableProperties(this._drawable, {visible: false});
+        }
+        this._renderPreviewFrame = null;
+    }
+
+    _setupPreview () {
+        const {renderer} = this.runtime;
+        if (!renderer) return;
```

Comment:
Return early if `this._video` is set to null. There is a race condition around `this._video` if video is quickly enabled and disabled.

Reply:
I'm not sure that this one matters, we'd just end up with a empty pen skin drawable ready for the video to toggle back on.

Reply:
It'll leave an empty pen skin but the next enable won't start the preview loop because _renderPreviewFrame is still set. Another disable followed by an extra enable would be needed.

Reply:
if `_renderPreviewFrame` is set - it is running... it just exits early if `!this.videoReady` -- so it will get called once per frame forever once we set it up, unless we remove it.

Reply:
Yes. You're right.